File: typescript-javascript/day06-errors-modules/exercises.js
// Day 6 exercises -- Errors and Modules. Run: node exercises.js

class NegativeAmountError extends Error {
    constructor(message) {
        super(message);
        this.name = "NegativeAmountError";
    }
}

class InsufficientFundsError extends Error {
    constructor(message) {
        super(message);
        this.name = "InsufficientFundsError";
    }
}

function safeDivide(a, b) {
    // Return a / b, but if b is 0, return null instead of Infinity/NaN.
    // (Dividing by zero in JS doesn't throw -- it gives Infinity or NaN --
    // so you'll need an explicit check here, not a try/catch.)
    // TODO: implement
}

function parseIntOrDefault(value, defaultValue = 0) {
    // Try to convert `value` (a string) to an integer using parseInt(value, 10).
    // If the result is NaN (Not a Number -- check with Number.isNaN()),
    // return defaultValue instead.
    // e.g. parseIntOrDefault("42") -> 42
    //      parseIntOrDefault("abc") -> 0
    // TODO: implement
}

function withdraw(balance, amount) {
    // Return balance - amount.
    // Throw a NegativeAmountError if amount is negative.
    // Throw an InsufficientFundsError if amount > balance.
    // TODO: implement, using `throw new ...Error(...)`
}

// ---------------------------------------------------------------------------
function check(label, condition) {
    console.log((condition ? "PASS" : "FAIL") + ": " + label);
}

check("safeDivide normal", safeDivide(10, 2) === 5);
check("safeDivide by zero", safeDivide(10, 0) === null);

check("parseIntOrDefault valid", parseIntOrDefault("42") === 42);
check("parseIntOrDefault invalid", parseIntOrDefault("abc") === 0);
check("parseIntOrDefault custom default", parseIntOrDefault("abc", -1) === -1);

check("withdraw normal", withdraw(100, 30) === 70);

try {
    withdraw(100, -5);
    check("withdraw throws on negative amount", false);
} catch (error) {
    check("withdraw throws on negative amount", error instanceof NegativeAmountError);
}

try {
    withdraw(100, 200);
    check("withdraw throws on insufficient funds", false);
} catch (error) {
    check("withdraw throws on insufficient funds", error instanceof InsufficientFundsError);
}

File: python/day06-errors-modules/exercises.py
"""Day 6 exercises — Errors, Modules. Run: python exercises.py"""

import math


class NegativeAmountError(Exception):
    """Custom exception for negative amounts."""
    pass


class InsufficientFundsError(Exception):
    """Custom exception for withdrawals exceeding balance."""
    pass


def safe_divide(a, b):
    """Return a / b, but if b is 0, return None instead of raising
    ZeroDivisionError. Use try/except, not an `if b == 0` check.
    """
    # TODO: implement
    pass


def parse_int_or_default(value, default=0):
    """Try to convert `value` (a string) to an int. If it can't be
    converted (raises ValueError), return `default` instead.
    e.g. parse_int_or_default("42") -> 42
         parse_int_or_default("abc") -> 0
    """
    # TODO: implement
    pass


def withdraw(balance, amount):
    """Return balance - amount.
    Raise NegativeAmountError if amount is negative.
    Raise InsufficientFundsError if amount > balance.
    (Both are the custom exceptions defined above.)
    """
    # TODO: implement, raising the appropriate custom exception
    pass


def sqrt_of_sum_of_squares(*numbers):
    """Return math.sqrt(sum of squares of numbers). Uses the `math` module
    imported at the top of this file -- practice using an imported module.
    e.g. sqrt_of_sum_of_squares(3, 4) -> 5.0
    """
    # TODO: implement
    pass


def main():
    """This should only print when the file is run directly, NOT when
    imported. (You don't need to change this function -- just make sure
    the `if __name__ == "__main__":` guard at the bottom is correct.)
    """
    print("Running Day 6 exercises directly.")


# ---------------------------------------------------------------------------
def check(label, condition):
    print(f"{'PASS' if condition else 'FAIL'}: {label}")


if __name__ == "__main__":
    main()

    check("safe_divide normal", safe_divide(10, 2) == 5.0)
    check("safe_divide by zero", safe_divide(10, 0) is None)

    check("parse_int_or_default valid", parse_int_or_default("42") == 42)
    check("parse_int_or_default invalid", parse_int_or_default("abc") == 0)
    check("parse_int_or_default custom default", parse_int_or_default("abc", -1) == -1)

    check("withdraw normal", withdraw(100, 30) == 70)

    try:
        withdraw(100, -5)
        check("withdraw raises on negative amount", False)
    except NegativeAmountError:
        check("withdraw raises on negative amount", True)

    try:
        withdraw(100, 200)
        check("withdraw raises on insufficient funds", False)
    except InsufficientFundsError:
        check("withdraw raises on insufficient funds", True)

    check("sqrt_of_sum_of_squares", sqrt_of_sum_of_squares(3, 4) == 5.0)
